scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1089,25 +1089,72 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
         case (pipe, cnt) => pipe.flatMap(List.fill(cnt)(_).iterator)
       }
       .map(_.toPipe[U](fieldNames)(flowDef, mode, setter)) // linter:ignore
-      .toList
+      .toVector
 
     if (merged.size == 1) {
       // there is no actual merging here, no need to rename:
       merged.head
     } else {
-      merged.reduce[Pipe] {
-        case (left, right) =>
-          // special handling for cases where one side of the hashjoin is merged
-          // with the hashjoin result. Cascading no longer allows it,
-          // so we add a checkpoint to the join result as a workaround
-          if (RichPipe.isHashJoinedWithPipe(left, right))
-            new Merge(RichPipe.assignName(new Checkpoint(left)), RichPipe.assignName(right))
-          else if (RichPipe.isHashJoinedWithPipe(right, left))
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(new Checkpoint(right)))
-          else
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(right))
-      }
+      val prepared  = mergeAvoidingHashes(merged)
+      prepared
+    }
+  }
+
+  /**
+    * Special handling for cases where one side of the hashjoin is merged
+    *  with the hashjoin result. Cascading no longer allows it (as of 3.0),
+    *  so we insert checkpoints and/or intermediate merge stages as appropriate
+    *
+    * @param pipes a list of pipes possibly colliding with each other
+    * @return an updated list of pipes, which can safely be merged
+    */
+  protected def mergeAvoidingHashes(pipes: Vector[Pipe]): Pipe = {
+    require(!pipes.isEmpty)
+
+    // we make use of the fact that the pipe merge operation is not just associative but also commutative
+    val (colliding, uncolliding) = pipes.partition(p => pipes.find(o => (o != p) && RichPipe.isHashJoinedWithPipe(p, o)).isDefined)
```

Comment:
Not sure. I initially needed recursion to deal with more complex examples, until I ended up splitting the ```case other @ (hj: HashJoin) =>``` case in  getChainOfEachs(). 
Now I need to verify whether the recursion still ever does anything (I suspect that now hash joins are transparent to getChainOfEachs(), we no longer really require it)

On your proposed change, I believe this would essentially generate my original code at https://github.com/twitter/scalding/pull/1605/commits/128b09513610f1b155839f3db5809e4a38bed61e, which @rubanm rejected on the grounds it would generate more Checkpoints than necessary. In the example from Monday, with ```a, b, c, d``` and ```a -> b, c -> d```, your code would output
```
new Merge(new Checkpoint(a), b, new Checkpoint(c), d)
```
which is necessary only if we also have ```a -> c```; from reading @rubanm, it seemed better to reach 
```
new Merge(new Checkpoint(new Merge(a,c)), b, d)
```
if a and c are hashjoinwise independent.